docs/proposals/kubelet-tls-bootstrap.md

```diff
@@ -0,0 +1,231 @@
+        IPAddresses []string
+
+        // Extra information the node wishes to send with the request
+        ExtraInfo []string
+}
+
+type CertificateSigningRequestStatus struct {
+        // Indicates whether CSR has a response yet. Default is Unknown. Status
+        // is True for approval and False for rejections.
+        Status api.ConditionStatus
+
+        // If CSR was rejected, these contain the reason why (if any was supplied).
+        Reason string
+        Message string
+
+        // If CSR was approved, this contains the issued certificate.
+        Certificate []byte
+}
+```
+
+We also introduce CertificateSigningRequestList to allow listing all the CSRs in the cluster:
+
+```go
+type CertificateSigningRequestList struct {
+        api.TypeMeta
+        api.ListMeta
+
+        Items []CertificateSigningRequest
+}
+```
+
+## Certificate Request Process
+
+### Node intialization
+
+When the kubelet executes it checks a location on disk for TLS assets
+(currently `/var/run/kubernetes/kubelet.{key,crt}` by default). If it finds
+them, it proceeds. If there are no TLS assets, the kubelet generates a keypair
+and self-signed certificate. We propose the following optional fallback behavior:
+
+1. Generate a keypair
+2. Generate a CSR for that keypair with CN set to the hostname (or
+   `--hostname-override` value) and DNS/IP SANs supplied with whatever values
+   the host knows for itself.
+3. Post the CSR to the CSR API endpoint.
+4. Set a watch on the CSR object to be notified of approval or rejection.
+
+### Controller response
+
+The apiserver must first validate the signature on the raw CSR data and reject
+requests featuring invalid CSRs. It then persists the
+CertificateSigningRequests and exposes the List of all CSRs for an
+administrator to approve or reject. The apiserver should watch for updates the
+Status field of any CertificateSigningRequest. When a CSR is approved
+(signified by Status changing from Unknown to True) the apiserver should
+generate and sign the certificate, then update the
+CertificateSigningRequestStatus with the new data.
+
+### Manual CSR approval
+
+An administrator using `kubectl` or another API client can query the
+CertificateSigningRequestList and update the status of
+CertificateSigningRequests. The default Status is Unknown, indicating that
+there has been no decision so far. A Status of True indicates that the admin
+has approved the request and the apiserver should issue the certificate. A
+Status of False indicates that the admin has denied the request. An admin may
+also supply Reason and Message fields to explain the rejection.
+
+## kube-apiserver support (CA assets)
+
+So that the apiserver can handle certificate issuance on its own, it will need
+access to CA signing assets. This could be as simple as a private key and a
+config file or as complex as a PKCS#11 client and supplementary policy system.
+For now, we will add flags for a signing key, a certificate, and a basic config
```

Comment:
I've been using cfssl in my prototyping work. They already define a policy file that supports all kinds of things we'll want in the future: https://github.com/cloudflare/cfssl/blob/master/config/config.go#L62